### PR TITLE
DS-4486 Add support/tests for Collection/Community role-based groups for Community/Collection Admins

### PR DESCRIPTION
**Pull request by Micheleboychuk**
## References
https://jira.lyrasis.org/browse/DS-4486

## Description
In this PR has been:
1. added another PreAuthorize check on EPersonRestRepository and GroupRestRepository search
    methods to enable access to community and collection administrators as well
2. created some IT to proof that the commuity/collection admins have access to the search  
    methods
3. created some IT to proof that the commuity/collection admins can manage their own groups 
    (submitters, workflow, etc.)
4. restore IT to proof that the community/collection admins can create/delete their own groups

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide)
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- N/A If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- N/A If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by abollini** [comment]
@Micheleboychuk thanks to have opened the PR today as planned. At the today dev meeting it was noted that another PR related to DS-4486 also exist, https://github.com/DSpace/DSpace/pull/2770
It includes ITs only, can you double check that all the scenario covered here are also included in your PR? if not, please include the missing test in this PR

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java [hunk 1]
```diff
@@ -654,6 +658,21 @@ public DSpaceObject getParentObject(Context context, Group group) throws SQLExce
                                 return collectionService.getParentObject(context, collection);
                             }
                         }
+                    } else {
+                        if (AuthorizeConfiguration.canCollectionAdminManagePolicies()) {
```
Comment (KevinVdV) [inline dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:662]: This if is in another if that checks the following:
`                if (AuthorizeConfiguration.canCollectionAdminManageWorkflows()
                        || AuthorizeConfiguration.canCommunityAdminManageCollectionWorkflows()) {
`
Should that if not be expanded to include this one ?
Should it also not a include a check on AuthorizeConfiguration.canCommunityAdminManagePolicies() ?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java [hunk 2]
```diff
@@ -140,14 +143,21 @@ private Authentication createAuthentication(final String password, final Context
 
         if (eperson != null) {
             boolean isAdmin = false;
+            boolean isCommunityAdmin = false;
+            boolean isColectionAdmin = false;
             try {
                 isAdmin = authorizeService.isAdmin(context, eperson);
+                isCommunityAdmin = authorizeService.isCommunityAdmin(context, eperson);
+                isColectionAdmin = authorizeService.isCollectionAdmin(context, eperson);
```
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java:151]: Small typo in the variable name, should be "isCollectionAdmin"

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java [hunk 3]
```diff
@@ -47,6 +48,8 @@
 
     private static final Logger log = LoggerFactory.getLogger(EPersonRestAuthenticationProvider.class);
 
+    public static final String ACCOUNT_ADMIN_GRANT = "ACCOUNT_ADMIN";
```
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java:51]: Can we rename this constant ? ACCOUNT_ADMIN_GRANT doesn't really point to the fact that the user could be a community / collection admin.

**Review by KevinVdV — CHANGES_REQUESTED** [review]
Made an initial code review, just a few small things to cleanup.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 4]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.not(Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        ))));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:489]: You don't need these two context lines here. You can change a configuration property in tests without turning off the authorization system.  Also, the configuration system resets itself after every test, so you may be able to remove all three lines here...you need not reset it back to the default value.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 5]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+                .content("https://localhost:8080/server/api/eperson/epersons/" + ePerson.getID()));
+
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        )));
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.admin-group", false);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:518]: Again, no need to turn off/on authorization here.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 6]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:532]: These 3 lines can be removed, no need to reset it to the default value, it'll do that automatically after the test runs.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 7]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.not(Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        ))));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:565]: Same here, can remove these 3 lines. It'll reset automatically.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 8]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.admin-group", false);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:595]: No need to turn off authorization to change a config value. remove context lines

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 9]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        token = getAuthToken(admin.getEmail(), password);
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:609]: These 3 lines can be removed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 10]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.not(Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        ))));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:704]: These 4 lines can be removed. configs will reset to defaults automatically.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 11]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:749]: These 4 lines can be removed. Configs reset automatically.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 12]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        )));
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", false);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", false);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:734]: No need to turn off authorization to set configs

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 13]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.not(Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        ))));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:784]: These 4 lines can be removed. Configs reset automatically.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 14]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:830]: These 4 lines can be removed. Configs reset automatically.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java [hunk 15]
```diff
@@ -437,4 +457,376 @@ public void deleteCommunityAdminGroupNotFoundTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", false);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", false);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityAdminGroupRestControllerIT.java:815]: No need to turn on/off authorization to set configs

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java [hunk 16]
```diff
@@ -1914,5 +1938,987 @@ public void patchNameTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:2840]: No need to reset these values. These 4 lines can be removed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java [hunk 17]
```diff
@@ -1914,5 +1938,987 @@ public void patchNameTest() throws Exception {
...
+
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/epersons"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.hasItem(
+                            EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
+                        )));
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", false);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", false);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:2825]: No need to turn off/on authorization to set configs. The context lines here can be removed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java [hunk 18]
```diff
@@ -1914,5 +1938,987 @@ public void patchNameTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.not(Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        ))));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:2875]: Again, these 4 lines can be removed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java [hunk 19]
```diff
@@ -1914,5 +1938,987 @@ public void patchNameTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", true);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", true);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:2921]: Same here, these 4 lines can be removed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java [hunk 20]
```diff
@@ -1914,5 +1938,987 @@ public void patchNameTest() throws Exception {
...
+        getClient(token).perform(get("/api/eperson/groups/" + adminGroup.getID() + "/subgroups"))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded.subgroups", Matchers.hasItem(
+                            GroupMatcher.matchGroupWithName(group.getName())
+                        )));
+
+        context.turnOffAuthorisationSystem();
+        configurationService.setProperty("core.authorization.community-admin.collection.admin-group", false);
+        configurationService.setProperty("core.authorization.collection-admin.admin-group", false);
+        context.restoreAuthSystemState();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:2906]: No need to turn on/off authorization to set configs.

**Review by tdonohue — COMMENTED** [review]
@Micheleboychuk and @abollini : I gave this a quick review.  Overall, it looks good.  But, I have a lot of minor suggestions inline below, most of them are the same mistake over and over again (which I'll add to our [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide)).  

Simply put, when you change configurations in a test,
1. You do **not** need to turn off the authorization system...the configurationService doesn't use the Context at all, and 
2. You do **not** need to reset the configs back to defaults, that will happen automatically in our testing framework. 

These are both tiny mistakes, but they make Integration tests a lot easier to read as you can remove 5-10 lines of code from each test.  No worries on the mistake, but I figured I'd note it where it occurs in this PR so you can clean it up.  As noted, I'll also update our Testing Guide as a reminder of how to change configs for testing.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java (hunk 3 above)
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java:51]: I agree, `MANAGE_ACCESS_GROUPS` looks indeed a better name

**Review by tdonohue — APPROVED** [review]
👍 This looks great to me now! Thanks for the quick updates @Micheleboychuk !

@KevinVdV : It looks to me like all your previous feedback was addressed.  Let me know if you'd like to do another quick review here or not.  I'll give you until end of day on Monday, and then I'll merge this as-is.  Thanks!

**Comment by tdonohue** [comment]
Merging, as this is at +1 and @KevinVdV's prior minor feedback has been addressed.

**Pull request merged**